integration/airflow/openlineage/airflow/listener.py

```diff
@@ -0,0 +1,141 @@
+import logging
+import threading
+import uuid
+import attr
+
+from typing import TYPE_CHECKING, Optional, Callable
+
+from airflow.listeners import hookimpl
+
+from openlineage.airflow.adapter import OpenLineageAdapter
+from openlineage.airflow.extractors import ExtractorManager
+from openlineage.airflow.utils import DagUtils, get_task_location, get_job_name, get_custom_facets
+
+if TYPE_CHECKING:
+    from airflow.models import TaskInstance, BaseOperator
+    from sqlalchemy.orm import Session
+
+
+@attr.s(frozen=True)
+class ActiveRun:
+    run_id: str = attr.ib()
+    task: "BaseOperator" = attr.ib()
+
+
+class ActiveRunManager:
+    """Class that stores run data - run_id and task in-memory. This is needed because Airflow
+    does not always pass all runtime info to on_task_instance_success and
+    on_task_instance_failed that is needed to emit events. This is not a big problem since
+    we're only running on worker - in separate process that is always spawned (or forked) on
+    execution, just like old PHP runtime model.
+    """
+    def __init__(self):
+        self.run_data = {}
+
+    def set_active_run(self, task_instance: "TaskInstance", run_id: str):
+        self.run_data[self._pk(task_instance)] = ActiveRun(run_id, task_instance.task)
+
+    def get_active_run(self, task_instance: "TaskInstance") -> Optional[ActiveRun]:
+        pk = self._pk(task_instance)
+        if pk not in self.run_data:
+            return None
+        return self.run_data[pk]
+
+    def _pk(self, ti: "TaskInstance"):
+        return ti.dag_id + ti.task_id + ti.run_id
+
+
+run_data_holder = ActiveRunManager()
+extractor_manager = ExtractorManager()
+adapter = OpenLineageAdapter()
+log = logging.getLogger('airflow')
+
+
+def execute_in_thread(target: Callable, kwargs=None):
+    if kwargs is None:
+        kwargs = {}
+    thread = threading.Thread(
+        target=target,
+        kwargs=kwargs,
+        daemon=True
+    )
+    thread.start()
+    # Join, but ignore checking if thread stopped. If it did, then we shoudn't do anything.
+    # This basically gives this thread 2 seconds to complete work, then it can be killed,
+    # as daemon=True. We don't want to deadlock Airflow if our code hangs.
+    thread.join(timeout=2)
+
+
+@hookimpl
+def on_task_instance_running(previous_state, task_instance: "TaskInstance", session: "Session"):
+    if not hasattr(task_instance, 'task'):
+        log.warning(f"No task set for TI object {id(task_instance)}")
```

Comment:
Wouldn't it be better to log `TI.task_id`, `TI.dag_id` and `TI.run_id` instead of  `id(task_instance)` ?

Reply:
Fixed.